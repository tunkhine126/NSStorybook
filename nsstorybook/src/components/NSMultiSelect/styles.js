import { makeStyles } from '@material-ui/core/styles';
import colors from '../../global-styles';

export const styles = (width = '100%', maxWidth) =>
  makeStyles({
    size: {
      width,
      maxWidth,
    },
    chips: {
      display: 'flex',
      flexWrap: 'wrap',
    },
    chip: {
      margin: 2,
    },
    chipLabelContainer: {
      fontSize: 12,
    },
    chipClearBtn: {
      marginTop: 1,
      marginLeft: 1,
      marginRight: 10,
    },
  });

export const useNSMultiSelectStyle = (
  width,
  height,
  padding = '27px 12px 10px 15px',
  minWidth
) =>
  makeStyles(theme => ({
    root: {
      backgroundColor: colors.INTERFACE.ui1,
      transition: theme.transitions.create(['border-color']),
      width,
      height,
      padding,
      minWidth,
    },
  }));
