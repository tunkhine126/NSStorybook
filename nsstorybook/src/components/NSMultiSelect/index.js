/**
 *
 * NSMultiSelect
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import OutlinedInput from '@material-ui/core/OutlinedInput';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import Chip from '@material-ui/core/Chip';

import colors from 'global-styles';
import { useNSMultiSelectStyle, styles } from './styles';
import PillClose from '../Icons/pillClose';

function NSMultiSelect({
  values,
  items,
  label,
  update,
  width,
  maxWidth = '400px',
  outterWidth,
  outerHeight,
  padding,
  disableLabel,
  selectOverride,
  placeholderText,
  required = false,
  translate = '14px, 10px',
  className,
  onBlur = () => {},
}) {
  const classes = styles(width, maxWidth)();
  const labelFormatted = label.replace(/ /g, '-').toLowerCase();

  const MenuProps = {
    PaperProps: {
      style: {
        maxHeight: 215,
      },
    },
  };

  const ComponentStyle = styled.div`
    .MuiInputLabel-outlined.MuiInputLabel-shrink {
      transform: translate(${translate}) scale(0.75);
      color: ${colors.STATE.active};
    }
    .MuiOutlinedInput-root:hover .MuiOutlinedInput-notchedOutline {
      border-color: ${colors.STATE.active};
    }
    .MuiOutlinedInput-notchedOutline {
      border-color: ${colors.INTERFACE.ui4};
    }
    .MuiSvgIcon-root {
      color: ${colors.TEXT.light};
    }
    .MuiInputLabel-outlined {
      color: ${colors.TEXT.light};
      width: 77%;
    }
    .MuiOutlinedInput-root {
      padding: 1px;
    }
    .MuiInputLabel-outlined {
      ${disableLabel ? 'display: none;' : ''}
    }
  `;

  let localValues = [];

  if (values.item) {
    localValues = !Array.isArray(values.item)
      ? values.item.split(' ')
      : values.item;
  } else {
    localValues = values;
  }

  return (
    <ComponentStyle>
      <FormControl
        variant="outlined"
        className={classes.size}
        required={required}
      >
        {!placeholderText && (
          <InputLabel htmlFor={`${labelFormatted}-selector`}>
            {label}
          </InputLabel>
        )}
        <Select
          multiple
          value={localValues}
          onChange={update}
          onBlur={onBlur}
          className={className}
          classes={selectOverride}
          displayEmpty={placeholderText}
          input={
            <OutlinedInput
              name={labelFormatted}
              id={`${labelFormatted}-selector`}
            />
          }
          renderValue={selected => (
            <div className={classes.chips}>
              {selected.map((v, idx) => (
                <Chip
                  key={idx}
                  id={`${v.toLowerCase()}-chip`}
                  label={
                    <span className={classes.chipLabelContainer}>
                      <span>{v}</span>
                    </span>
                  }
                  onDelete={() => {
                    update(selected.filter(item => item !== v));
                  }}
                  deleteIcon={<PillClose className={classes.chipClearBtn} />}
                  color="primary"
                  className={classes.chip}
                />
              ))}
            </div>
          )}
          inputProps={{
            classes: useNSMultiSelectStyle(outterWidth, outerHeight, padding)(),
          }}
          MenuProps={MenuProps}
        >
          {items.map((item, idx) => (
            <MenuItem key={idx} value={item.value}>
              {item.value}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    </ComponentStyle>
  );
}

NSMultiSelect.propTypes = {
  values: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  items: PropTypes.array.isRequired,
  update: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  className: PropTypes.string,
  onBlur: PropTypes.func,
  required: PropTypes.bool,
  placeholderText: PropTypes.bool,
  disableLabel: PropTypes.bool,
  width: PropTypes.string,
  maxWidth: PropTypes.string,
  outterWidth: PropTypes.string,
  outerHeight: PropTypes.string,
  padding: PropTypes.string,
  translate: PropTypes.string,
  selectOverride: PropTypes.object,
};

export default NSMultiSelect;
