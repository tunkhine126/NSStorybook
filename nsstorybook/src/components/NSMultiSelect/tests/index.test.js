/**
 *
 * Tests for NSMultiSelect
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from '@testing-library/react';

import NSMultiSelect from '../index';

describe('<NSMultiSelect />', () => {
  const colors = ['red', 'green', 'blue'];

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <NSMultiSelect
        values={{ item: [] }}
        items={colors}
        update={() => {}}
        label="message"
      />
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <NSMultiSelect
        values={{ item: [] }}
        items={colors}
        update={() => {}}
        label="Message"
      />
    );
    expect(firstChild).toMatchSnapshot();
  });
});
