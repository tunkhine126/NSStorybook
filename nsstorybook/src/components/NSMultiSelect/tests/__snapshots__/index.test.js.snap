// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NSMultiSelect /> Should render and match the snapshot 1`] = `
<div
  class="NSMultiSelect__ComponentStyle-sc-1sbmilg-0 curCdr"
>
  <div
    class="MuiFormControl-root makeStyles-size-91"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink="false"
      for="message-selector"
    >
      Message
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby=" mui-component-select-message"
        class="MuiSelect-root makeStyles-root-96 MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        id="mui-component-select-message"
        role="button"
        tabindex="0"
      >
        <span>
          ​
        </span>
      </div>
      <input
        id="message-selector"
        name="message"
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-177 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-178"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
